### DISCUSSION TITLE
IDE Support/Workflow

### DISCUSSION THREAD

**Discussion by JulianBissekkou** [leancodepl/patrol#1233] [Q&A]
I would like to know how your workflow is. I really enjoy starting my tests from IntelliJ and viewing the results without touching the command line.

How do you debug and run the tests in your IDE, or are you using the command line for most things?

**Comment by bartekpacia** [comment]
Hi,

We use terminal for everything (`patrol develop` for debugging tests, `patrol test` on CI). `adb logcat` also helps in rare and annoying cases.

On CI, we take advantage of the native results (JUnit XML on Android, `.xcresult` on iOS). Some providers, like Bitrise, have really nice integrations with it. 

But our integration with native testing tools a bit barebones now (biggest pain is that we get to run only 1 test in a single APK. To run another test, a new APK has to be built). This will get 10x better when we finish the [epic test bundling feature](https://github.com/leancodepl/patrol/issues/1004).

I agree that starting tests from IJ is cool. Unfortunately, we don't support it (technically we do, because we're just a JUnit or XCTest, but `patrol_cli` performs more things that are required to make Patrol work). In other words, `patrol test` and clicking the green "Run tests" button is not equivalent. `patrol test` does a superset actions that happen when you click "Run tests" in your IDE (be it IJ or Xcode).

**Reply by JulianBissekkou** [reply]
A "low" effort solution would be to give users the ability to create run configurations and maybe we can "simply" attach a debugger to the command that is executed.
It's totally fine to create a runconfig which executes `patrol test --testname="some test name" --attach-debugger`
Attaching the debugger might work by just calling `flutter attach`? Not sure because I haven't digged into that part yet.

It we get this working its only a small step to create a vscode or intellij plugin that starts test annotated with `@patroltest` from the line of code. (using the "▶️" icon as you do now with pure java or dart tests)

**Reply by bartekpacia** [reply]
This sounds cool.

We'd probably add the hypothetical `--attach-debugger` flag to `patrol develop`, not `patrol test`.

`flutter attach` sounds like _the_ command to do it. We'd have to look into how IDE extensions (e.g. Dart Code) attach to a running app. I see Dart Code has the command to attach to a running app:

![Screenshot 2023-05-09 at 3 42 47 PM](https://github.com/leancodepl/patrol/assets/40357511/12c7818e-5462-4c76-ae20-afb70d01f0b9)

After clicking Enter, it asks for the Dart VM service address:

![Screenshot 2023-05-09 at 3 43 03 PM](https://github.com/leancodepl/patrol/assets/40357511/bfd133b9-e720-45a7-af94-d7428d931235)

The VM service address is visible during `flutter run`:

```console
$ flutter run
Using hardware rendering with device sdk gphone64 arm64. If you notice graphics artifacts, consider enabling software rendering with "--enable-software-rendering".
Launching lib/main.dart on sdk gphone64 arm64 in debug mode...
Running Gradle task 'assembleDebug'...                              3.6s
✓  Built build/app/outputs/flutter-apk/app-debug.apk.
D/FlutterGeolocator(23828): Attaching Geolocator to activity
D/FlutterGeolocator(23828): Creating service.
D/FlutterGeolocator(23828): Binding to location service.
D/FlutterGeolocator(23828): Geolocator foreground service connected
D/FlutterGeolocator(23828): Initializing Geolocator services
D/FlutterGeolocator(23828): Flutter engine connected. Connected engine count 1
Syncing files to device sdk gphone64 arm64...                       72ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

A Dart VM Service on sdk gphone64 arm64 is available at: http://127.0.0.1:57616/umaqEOhVP28=/
The Flutter DevTools debugger and profiler on sdk gphone64 arm64 is available at: http://127.0.0.1:9101?uri=http://127.0.0.1:57616/umaqEOhVP28=/
```

URI is `http://127.0.0.1:57616/umaqEOhVP28=/`. But after entering it here and clicking Enter:

![Screenshot 2023-05-09 at 3 47 44 PM](https://github.com/leancodepl/patrol/assets/40357511/9240ce8f-de0d-462b-a56a-e9bcc8838168)

... I get this infinitely:

![Screenshot 2023-05-09 at 3 46 23 PM](https://github.com/leancodepl/patrol/assets/40357511/325923a4-de5b-4d32-acb4-4ea6878e3eaa)

Maybe I'm doing something wrong?
Anyway, it does seem like _a way_ to add debugger support for tests.

FWIW, clicking the second link ("The Flutter DevTools debugger and profiler") works. It opens the DevTools in the browser.


@mateuszwojtczak do you have any ideas?

**Reply by JulianBissekkou** [reply]
I will give this a round of testing tomorrow, but I would love to contribute to the patrol intelliJ plugin to make things work seemless ❤️

**Reply by bartekpacia** [reply]
Wow, that'd be so awesome! 

We'd have to be more wary of breaking changes under the hood, though :) This was a problem e.g. [here](https://github.com/leancodepl/patrol/discussions/1180).

**Comment by adrian-moisa** [comment]
Any progress made on attaching a debugger? Would be extremely useful. I've been able to debug without the debugger but I had quite some tough nuts to crack early on. Now things are stable. Though still I'd be super happy to have access to a debugger. Cheers!
